src/main/java/org/jenkinsci/plugins/pipeline/milestone/MilestoneStepExecution.java

```diff
@@ -281,80 +168,83 @@ private static void println(StepContext context, String message) {
-                Integer number = it2.next();
-                if (job.getBuildByNumber(number) == null) {
-                    // Deleted at some point but did not properly clean up from exit(…).
-                    LOGGER.log(WARNING, "Cleaning up apparently deleted {0}#{1}", new Object[] {jobName, number});
-                    it2.remove();
+        @Override public void onCompleted(Run<?,?> r, @NonNull TaskListener listener) {
+            if (isPipelineRun(r)) {
+                MilestoneStorage milestoneStorage = getStorage();
+                var result = milestoneStorage.clear(r);
+                MilestoneStorage.LOGGER.finest(() -> "milestones after completion: " + result.milestones());
+                if (result.lastMilestoneBeforeCompletion() != null) {
+                    MilestoneStorage.LOGGER.finest(() -> "Build" + r + " last milestone before completion: " + result.lastMilestoneBeforeCompletion());
+                    var buildsToCancel = getBuildsToCancel(r.getNumber(), Integer.MAX_VALUE, result.milestones());
+                    milestoneStorage.cancel(r.getParent(), buildsToCancel);
+                } else {
+                    MilestoneStorage.LOGGER.finest(() -> "Build " + r + " was not using milestones, nothing to cancel");
                 }
             }
         }
-    }
 
-    private static void save() {
-        Jenkins.get().getDescriptorOrDie(MilestoneStep.class).save();
+        private boolean isPipelineRun(Run<?, ?> r) {
+            return r instanceof FlowExecutionOwner.Executable executable && executable.asFlowExecutionOwner() != null;
+        }
     }
 
+
     @Extension
-    public static final class Listener extends RunListener<Run<?,?>> {
-        @Override public void onCompleted(Run<?,?> r, TaskListener listener) {
-            if (!(r instanceof FlowExecutionOwner.Executable) || ((FlowExecutionOwner.Executable) r).asFlowExecutionOwner() == null) {
-                return;
+    public static final class ItemListenerImpl extends ItemListener {
+        @Override
+        public void onDeleted(Item item) {
+            if (item instanceof Job<?,?> job) {
+                getStorage().onDeletedJob(job);
             }
-            exit(r);
         }
     }
 
     /**
-     * Clean up tracked jobs on deleted.
+     * Listens to pipeline resume, and let {@link MilestoneStorage} know the latest persisted milestone.
      */
     @Extension
-    public static class CleanupJobsOnDelete extends ItemListener {
-
+    public static final class FlowExecutionListenerImpl extends FlowExecutionListener {
         @Override
-        public void onDeleted(Item item) {
-            if (item instanceof Job) {
-                String jobName = item.getFullName();
-                Map<Integer, Milestone> job = getMilestonesByOrdinalByJob().get(jobName);
-                if (job != null) {
-                    remove(jobName);
+        public void onResumed(@NonNull FlowExecution execution) {
```

Comment:
Remember that this will happen after startup, and in a nondeterministic order relative to other builds. Are there possible race conditions there?

Reply:
https://github.com/jenkinsci/pipeline-milestone-step-plugin/pull/147/files#r1968108696 I guess but this is already handled theoretically. I don't see another case.